community/modules/scheduler/pbspro-server/variables.tf

```diff
@@ -0,0 +1,410 @@
+  description = "The self link of the subnetwork to attach the VM."
+  type        = string
+  default     = null
+}
+
+variable "network_interfaces" {
+  description = <<-EOT
+    A list of network interfaces. The options match that of the terraform
+    network_interface block of google_compute_instance. For descriptions of the
+    subfields or more information see the documentation:
+    https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_instance#nested_network_interface
+
+    **_NOTE:_** If `network_interfaces` are set, `network_self_link` and
+    `subnetwork_self_link` will be ignored, even if they are provided through
+    the `use` field. `bandwidth_tier` and `disable_public_ips` also do not apply
+    to network interfaces defined in this variable.
+
+    Subfields:
+    network            (string, required if subnetwork is not supplied)
+    subnetwork         (string, required if network is not supplied)
+    subnetwork_project (string, optional)
+    network_ip         (string, optional)
+    nic_type           (string, optional, choose from ["GVNIC", "VIRTIO_NET"])
+    stack_type         (string, optional, choose from ["IPV4_ONLY", "IPV4_IPV6"])
+    queue_count        (number, optional)
+    access_config      (object, optional)
+    ipv6_access_config (object, optional)
+    alias_ip_range     (list(object), optional)
+    EOT
+  type = list(object({
+    network            = string,
+    subnetwork         = string,
+    subnetwork_project = string,
+    network_ip         = string,
+    nic_type           = string,
+    stack_type         = string,
+    queue_count        = number,
+    access_config = list(object({
+      nat_ip                 = string,
+      public_ptr_domain_name = string,
+      network_tier           = string
+    })),
+    ipv6_access_config = list(object({
+      public_ptr_domain_name = string,
+      network_tier           = string
+    })),
+    alias_ip_range = list(object({
+      ip_cidr_range         = string,
+      subnetwork_range_name = string
+    }))
+  }))
+  default = []
+  validation {
+    condition = alltrue([
+      for ni in var.network_interfaces : (ni.network == null) != (ni.subnetwork == null)
+    ])
+    error_message = "All additional network interfaces must define exactly one of \"network\" or \"subnetwork\"."
+  }
+  validation {
+    condition = alltrue([
+      for ni in var.network_interfaces : ni.nic_type == "GVNIC" || ni.nic_type == "VIRTIO_NET" || ni.nic_type == null
+    ])
+    error_message = "In the variable network_interfaces, field \"nic_type\" must be either \"GVNIC\", \"VIRTIO_NET\" or null."
+  }
+  validation {
+    condition = alltrue([
+      for ni in var.network_interfaces : ni.stack_type == "IPV4_ONLY" || ni.stack_type == "IPV4_IPV6" || ni.stack_type == null
+    ])
+    error_message = "In the variable network_interfaces, field \"stack_type\" must be either \"IPV4_ONLY\", \"IPV4_IPV6\" or null."
+  }
+}
```

Comment:
nit: Why not just accept object(any) and let type enforcement be done but sub-module. Documentation would take a hit but you could just put a link to vm-instance documentation. I don't feel to strongly about this as there is an obvious down side with documentation.

Reply:
That position is reasonable.

Since wrapping Terraform modules is not a simple matter as `x inherits y` I have leaned toward literal duplication of code as the mechanism. I also feel like setting `any` within a variable type is something does one has no other options ("last refuge of...").

You will see an example of being compelled to use `any` in the final PR of this series.

"ACK"